updater/internal/state/monitor.go

```diff
@@ -0,0 +1,139 @@
+	// It will return an error if status is Failed or if the context times out.
+	MonitorForSuccess(ctx context.Context, packageName string) error
+}
+
+// CollectorMonitor implements Monitor interface for monitoring the Collector Package Status file
+type CollectorMonitor struct {
+	stateManager  packagestate.StateManager
+	currentStatus *protobufs.PackageStatuses
+}
+
+// NewCollectorMonitor create a new Monitor specifically for the collector
+func NewCollectorMonitor(logger *zap.Logger) (Monitor, error) {
+	// Get install directory
+	installDir, err := path.InstallDir()
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine install directory: %w", err)
+	}
+
+	// Create a collector monitor
+	packageStatusPath := filepath.Join(installDir, packagestate.DefaultFileName)
+	collectorMonitor := &CollectorMonitor{
+		stateManager: packagestate.NewFileStateManager(logger, packageStatusPath),
+	}
+
+	// Load the current status to ensure the package status file exists
+	collectorMonitor.currentStatus, err = collectorMonitor.stateManager.LoadStatuses()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load package statues: %w", err)
+	}
+
+	return collectorMonitor, nil
+
+}
+
+// SetState sets the status on the specified package and saves it to the package status file
+func (c *CollectorMonitor) SetState(packageName string, status protobufs.PackageStatus_Status, statusErr error) error {
+	// Verify we have package by that name
+	targetPackage, ok := c.currentStatus.GetPackages()[packageName]
+	if !ok {
+		return fmt.Errorf("no package for name %s", packageName)
+	}
+
+	// Update the status
+	targetPackage.Status = status
+
+	// If that passed in error is not nil set it as the error message
+	if statusErr != nil {
+		targetPackage.ErrorMessage = statusErr.Error()
+	}
+
+	c.currentStatus.GetPackages()[packageName] = targetPackage
+
+	// Save to updated status to disk
+	return c.stateManager.SaveStatuses(c.currentStatus)
+}
+
+// MonitorForSuccess intermittently checks the package status file for either an install failed or success status.
+// If an InstallFailed status is read this returns ErrFailedStatus error.
+// If the context is canceled or there is an error reading the status file the error will be returned.
+func (c *CollectorMonitor) MonitorForSuccess(ctx context.Context, packageName string) error {
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			packageStatus, err := c.stateManager.LoadStatuses()
+			switch {
+			case errors.Is(err, os.ErrNotExist):
+				// File may not exist for a short period of time while being written so just continue
+				continue
+			case err != nil:
+				return fmt.Errorf("error while loaded status: %w", err)
+			default:
+				targetPackage, ok := packageStatus.GetPackages()[packageName]
+				// Target package might not exist yet so continue
```

Comment:
@cpheps how could the target package not exist yet in the status file?

Reply:
This is kind of future proofing. This shouldn't happen but I'm protecting against it just in case.

Reply:
i mean, i guess it doesn't hurt anything right now so ok

Reply:
Yeah, mostly I had to do a map lookup and I didn't want to assume it would just be ok. I though retrying then hitting a timeout was more flexible than just failing.